components/scream/cime_config/buildnml

```diff
@@ -120,9 +111,115 @@ def evaluate_selectors(element, case, ez_selectors):
+    True
+    >>> get_child(good,'var4').text=="bar_off"
+    True
+    >>> ############## BAD SELECTOR DEFINITION #####################
+    >>> xml_sel_bad1 = '''
+    ... <selectors_xml>
+    ...   <selectors>
+    ...     <selector name="grid" case_env="BADENV"/>
+    ...   </selectors>
+    ... </selectors_xml>
+    ... '''
+    >>> selectors_bad1 = get_valid_selectors(ET.fromstring(xml_sel_bad1))
+    >>> good = ET.fromstring(xml_good)
+    >>> evaluate_selectors(good,case,selectors_bad1)
+    Traceback (most recent call last):
+    evaluate_selectors.<locals>.UnknownCaseValue: ERROR: Bad easy selector 'grid' definition. Relies on unknown case value 'BADENV'
+    >>> ############## BAD SELECTOR DEFINITION #####################
+    >>> xml_sel_bad2 = '''
+    ... <selectors_xml>
+    ...   <selectors>
+    ...     <selector name="grid" case_env="ATM_GRID" regex=".*"/>
+    ...   </selectors>
+    ... </selectors_xml>
+    ... '''
+    >>> selectors_bad2 = get_valid_selectors(ET.fromstring(xml_sel_bad2))
+    >>> good = ET.fromstring(xml_good)
+    >>> evaluate_selectors(good,case,selectors_bad2)
+    Traceback (most recent call last):
+    evaluate_selectors.<locals>.MultipleMatches: ERROR: Selector 'grid' has invalid custom regex '.*' which does not capture exactly 1 group
+    >>> ############## BAD SELECTOR NAME #####################
+    >>> xml_bad1 = '''
+    ... <namelist_defaults>
+    ...   <var1>zero</var1>
+    ...   <var1 my_grid="ne4ne4">one</var1>
+    ... </namelist_defaults>
+    ... '''
+    >>> bad1 = ET.fromstring(xml_bad1)
+    >>> evaluate_selectors(bad1,case,selectors_good)
+    Traceback (most recent call last):
+    evaluate_selectors.<locals>.BadSelectorName: ERROR: Bad selector 'my_grid' for child 'var1'. 'my_grid' is not a valid case value or easy selector
+    >>> ############## BAD DEFAULTS ORDERING #####################
+    >>> xml_bad2 = '''
+    ... <namelist_defaults>
+    ...   <var1 grid="ne4ne4">one</var1>
+    ...   <var1>zero</var1>
+    ... </namelist_defaults>
+    ... '''
+    >>> bad2 = ET.fromstring(xml_bad2)
+    >>> evaluate_selectors(bad2,case,selectors_good)
+    Traceback (most recent call last):
+    evaluate_selectors.<locals>.BadDefaultsOrdering: ERROR: child 'var1' element without selectors occured after other parameter elements for this parameter
+    >>> ############## MULTIPLE MATCHES #####################
+    >>> xml_bad3 = '''
+    ... <namelist_defaults>
+    ...   <var1>zero</var1>
+    ...   <var1>one</var1>
+    ... </namelist_defaults>
+    ... '''
+    >>> bad3 = ET.fromstring(xml_bad3)
+    >>> evaluate_selectors(bad3,case,selectors_good)
+    Traceback (most recent call last):
+    evaluate_selectors.<locals>.BadDefaultsOrdering: ERROR: child 'var1' element without selectors occured after other parameter elements for this parameter
     """
-    metadata = {} # elem_name -> (type, valid_values)
+
+    class BadSelectorName (Exception):
```

Comment:
Do we really need custom exception types? This is only useful if we are catching and handling multiple types of exceptions, e.g.

```
try:
    * some action *
except CustomExc1:
    * handle exc1 error condition *
except CustomExc2:
    * handle exc2 error condition *
```

We aren't doing that anywhere as far as I've looked, so the custom exceptions are not necessary. Checking the exception string in doctest is adequate for ensuring the expected error was hit.

Another problem is that these new exceptions are not SystemExit exception and so will give the user an ugly stacktrace that is likely useless to them. They only need to see the msg describing the XML mistake they made.